Detectors/TPC/calibration/src/CalibdEdx.cxx

```diff
@@ -11,91 +11,207 @@
 
 #include "TPCCalibration/CalibdEdx.h"
 
+#include <algorithm>
 #include <array>
+#include <cmath>
 #include <cstddef>
-#include <memory>
+#include <gsl/span>
+#include <numeric>
 #include <string_view>
+#include <utility>
 
 //o2 includes
-#include "CommonUtils/MemFileHelper.h"
-#include "CommonUtils/TreeStreamRedirector.h"
-#include "CCDB/CcdbApi.h"
-#include "DetectorsCalibration/Utils.h"
+#include "DataFormatsTPC/TrackTPC.h"
+#include "DataFormatsTPC/TrackCuts.h"
+#include "DataFormatsTPC/Defs.h"
 #include "Framework/Logger.h"
-#include "TPCCalibration/dEdxHistos.h"
+#include "MathUtils/fit.h"
+#include "MathUtils/Utils.h"
+#include "TPCCalibration/FastHisto.h"
+
+//root includes
+#include "TFile.h"
+#include "TH2F.h"
+
+// boost includes
+#include <boost/histogram.hpp>
 
 using namespace o2::tpc;
+namespace bh = boost::histogram;
 
-void CalibdEdx::initOutput()
+CalibdEdx::CalibdEdx(int nBins, float minTotdEdx, float maxTotdEdx,
+                     float minMaxdEdx, float maxMaxdEdx, const TrackCuts& cuts)
+  : mCuts{cuts}, mNBins{nBins}
 {
-  // Here we initialize the vector of our output objects
-  mInfoVector.clear();
-  mMIPVector.clear();
-  return;
+  mTotHist = bh::make_histogram(
+    bh::axis::regular<>(nBins, minTotdEdx, maxTotdEdx, "dEdx"),
+    HistIntAxis(0, SECTORSPERSIDE, "sector"),
+    HistIntAxis(0, SIDES, "side"),
+    HistIntAxis(0, GEMSTACKSPERSECTOR, "type"));
+
+  mMaxHist = bh::make_histogram(
+    bh::axis::regular<>(nBins, minMaxdEdx, maxMaxdEdx, "dEdx"),
+    HistIntAxis(0, SECTORSPERSIDE, "sector"),
+    HistIntAxis(0, SIDES, "side"),
+    HistIntAxis(0, GEMSTACKSPERSECTOR, "type"));
 }
 
-void CalibdEdx::finalizeSlot(Slot& slot)
+void CalibdEdx::fill(const TrackTPC& track)
 {
-  LOG(INFO) << "Finalizing slot " << slot.getTFStart() << " <= TF <= " << slot.getTFEnd();
+  // applying cuts
+  if (track.hasBothSidesClusters() || (mApplyCuts && !mCuts.goodTrack(track))) {
+    return;
+  }
 
-  const dEdxHistos* container = slot.getContainer();
-  const auto statsASide = container->getHists()[0].getStatisticsData();
-  const auto statsCSide = container->getHists()[1].getStatisticsData();
+  const auto& dEdx = track.getdEdx();
+  const auto side = track.hasASideClustersOnly() ? Side::A : Side::C;
+
+  for (const GEMstack stack : {IROCgem, OROC1gem, OROC2gem, OROC3gem}) {
+    bool ok = false;
+    const int sector = findTrackSector(track, stack, ok);
+
+    // Ignore stack if we are not able to find its sector
+    if (!ok) {
+      continue;
+    }
+
+    // Fill the correct readout type
+    switch (stack) {
```

Comment:
You could consider getting rid of the switch by adding before the `for` loop:
```
std::vector<float> dEdxTot{dEdx.dEdxTotIROC, dEdx.dEdxTotOROC1, dEdx.dEdxTotOROC2, dEdx.dEdxTotOROC3};
// same for Max
```
And here simply do
```
mTotHist(dEdxTot[stack], sector, side, stack);
// same for Max
```